File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/service/UpdateWorker.java
package com.byagowi.persiancalendar.service;

import android.content.Context;

import com.byagowi.persiancalendar.utils.UpdateUtils;
import com.byagowi.persiancalendar.utils.Utils;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

public class UpdateWorker extends Worker {
    public UpdateWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Worker.Result doWork() {
        Utils.setChangeDateWorker();
        Utils.updateStoredPreference(getApplicationContext());
        UpdateUtils.update(getApplicationContext(), true);
        return Result.success();
    }
}

File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/service/UpdateWorker.kt
package com.byagowi.persiancalendar.service

import android.content.Context
import android.util.Log
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import com.byagowi.persiancalendar.utils.setChangeDateWorker
import com.byagowi.persiancalendar.utils.update
import com.byagowi.persiancalendar.utils.updateStoredPreference
import kotlinx.coroutines.coroutineScope

const val TAG = "UpdateWorker"

class UpdateWorker(context: Context, workerParams: WorkerParameters) :
        CoroutineWorker(context, workerParams) {
    override suspend fun doWork(): Result = coroutineScope {
        try {
            setChangeDateWorker(applicationContext)
            updateStoredPreference(applicationContext)
            update(applicationContext, true)
            Result.success()
        } catch (e: Throwable) {
            Log.e(TAG, "Error running UpdateWorker: ", e)
            Result.failure()
        }
    }
}
